### PR TITLE
Make dataloader use another random generator

### PR DESCRIPTION
**Pull request by thomasw21**
This is in order to start synchronizing the dropout across TP.

### CODE REVIEW

**Comment by thomasw21** [comment]
1. Yes, we don't actually care about the seed of each worker since data loading should not be random. 

The issue was `base_seed is a long generated by main process using its RNG (thereby, consuming a RNG state mandatorily)` Which means that tp_rank=0 would consume a a RNG state, while other ranks wouldn't.

**Comment by stas00** [comment]
> tp_rank=0 would consume a a RNG state, while other ranks wouldn't.

why won't the other ranks?

I think here by "main process" it refers to the process that spawns the dataloader workers. Won't that main process be each rank's process with relationship to its workers? i.e. there are multiple main processes in this context.

Since I haven't investigated this - have you validated that this fix actually changes something? I assume you do, but you haven't described this in the OP, hence the asking.

**Comment by thomasw21** [comment]
> why won't the other ranks?
https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/87a9dba0b9e91c09c410229f2ab97dec9c98afe0/megatron/training.py#L1170-L1174 This line only runs on `tp_rank = 0` if you go up the file a bit. The main process here will refer to `tp_rank=0`

> Since I haven't investigated this - have you validated that this fix actually changes something? 
Yes I was able to reproduce the discrepancy in `torch_rng_state` accross tp ranks, and this fixed it.

> I assume you do, but you haven't described this in the OP, hence the asking.
My bad, since we talked about this on slack I didn't copy paste the thread.

**Review by stas00 — APPROVED** [review]
Thank you for adding the missing sync, Thomas
